torchgeo/datasets/geo.py

```diff
@@ -310,7 +311,11 @@ def files(self) -> list[Path]:
             if os.path.isdir(path):
                 pathname = os.path.join(path, '**', self.filename_glob)
                 files |= set(glob.iglob(pathname, recursive=True))
-            elif os.path.isfile(path) or path_is_vsi(path):
+            elif (
+                os.path.isfile(path)
+                and fnmatch.fnmatch(str(path), os.path.join('*', self.filename_glob))
```

Comment:
Alternative: use `fnmatch.filter` at the end. This won't result in a warning though. This could be a good thing or a bad thing.

Reply:
https://github.com/python/typeshed/pull/12522